src/main/kotlin/org/rust/lang/core/completion/RsImplTraitMemberCompletionProvider.kt

```diff
@@ -0,0 +1,127 @@
+        if (implBlock.traitRef == null) return
+
+        val trait = implBlock.implementedTrait ?: return
+        val members = trait.element.members ?: return
+
+        val parentItems = members.expandedMembers.toMutableSet()
+        for (item in implBlock.expandedMembers) {
+            parentItems.removeIf { it.javaClass == item.javaClass && it.name == item.name }
+        }
+
+        for (item in parentItems) {
+            val lookup = getCompletion(item, trait, implBlock) ?: continue
+            result.addElement(
+                lookup.withPriority(KEYWORD_PRIORITY + 1)
+            )
+        }
+    }
+}
+
+private fun getCompletion(
+    target: RsAbstractable,
+    trait: BoundElement<RsTraitItem>,
+    impl: RsImplItem
+): LookupElementBuilder? {
+    val substitution = trait.subst
+    return when (target) {
+        is RsConstant -> completeConstant(target, substitution)
+        is RsTypeAlias -> completeType(target)
+        is RsFunction -> completeFunction(target, trait, impl)
+        else -> error("unreachable")
+    }
+}
+
+private fun completeConstant(target: RsConstant, substitution: Substitution): LookupElementBuilder? {
+    val text = buildString {
+        append("const ")
+        append(target.name)
+        val type = target.typeReference?.type?.substitute(substitution) ?: return null
+        append(": ")
+        append(type.renderInsertionSafe())
+        val builder = RsDefaultValueBuilder(target.knownItems, target.containingMod, RsPsiFactory(target.project))
+        append(" = ")
+        append(builder.buildFor(type, mapOf()).text)
+        append(';')
+    }
+    return LookupElementBuilder.create(text)
+        .withIcon(RsIcons.CONSTANT)
+        .withInsertHandler { context, _ ->
+            val element = context.getElementOfType<RsConstant>() ?: return@withInsertHandler
+            val expr = element.expr ?: return@withInsertHandler
+            selectElement(expr, context.editor)
+        }
+}
+
+private fun completeType(target: RsTypeAlias): LookupElementBuilder {
+    val text = buildString {
+        append("type ")
+        append(target.name)
+        append(" = ();")
+    }
+    return LookupElementBuilder.create(text)
+        .withIcon(RsIcons.TYPE_ALIAS)
+        .withInsertHandler { context, _ ->
+            val element = context.getElementOfType<RsTypeAlias>() ?: return@withInsertHandler
+            val typeReference = element.typeReference ?: return@withInsertHandler
+            selectElement(typeReference, context.editor)
+        }
+}
+
+private fun completeFunction(
+    target: RsFunction,
+    trait: BoundElement<RsTraitItem>,
+    impl: RsImplItem
+): LookupElementBuilder {
+    val substitution = trait.subst
+    val signature = target.getSignatureText(substitution, trait, impl) ?: ""
+    val text = "$signature{\n        unimplemented!()\n    }"
+    return LookupElementBuilder
+        .create(text)
+        .withIcon(RsIcons.FUNCTION)
```

Comment:
Probably, we need to reuse `org.rust.lang.core.psi.ext.RsFunctionImplMixin#getIcon` and its friends in `RsConstantImplMixin` and `RsTypeAliasImplMixin` somehow (I suppose we need to extract the corresponding code into the corresponding util methods) to use the same icon style

Reply:
Maybe we can keep it simpler, because we know that the completion will only be in a certain situation?
The owner will always be a trait, the abstractable will therefore always be public and non-abstract. Therefore it should be `RsIcons.ASSOC_TYPE_ALIAS` for type aliases, `RsIcons.ASSOC_CONSTANT`  for constants and functions should be either `RsIcons.ASSOC_FUNCTION` or `RsIcons.METHOD`.

But if you want, I can also extract these methods. What should I use as the flags in that case? `Iconable.ICON_FLAG_VISIBILITY | Iconable.ICON_FLAG_READ_STATUS`?

Reply:
> Maybe we can keep it simpler, because we know that the completion will only be in a certain situation?
The owner will always be a trait, the abstractable will therefore always be public and non-abstract. 

Ideally, I'd prefer a single place for such logic because if we decide to change something related to icons in the future, it would be great to modify only single place

> But if you want, I can also extract these methods. What should I use as the flags in that case? Iconable.ICON_FLAG_VISIBILITY | Iconable.ICON_FLAG_READ_STATUS?

We use `0` during completion (see `org.rust.lang.core.completion.LookupElementsKt#getLookupElementBuilder`) so I suppose it's ok to use `0` here as well

Reply:
It's straightforward to use the icon of the parent item (https://github.com/intellij-rust/intellij-rust/pull/6342/commits/08aecd6a413fb5fc5957ec1bb44b5c019668d444). I'm not sure if it's better to show the parent or the completed item icon.

But for the latter, to reuse `getIcon` properly, we would probably need to construct a fake PSI element on the completed location and then get it's icon. It sounds hacky. Or we would need to refactor `getIcon` so that it doesn't need an existing PSI element.

Reply:
Yeah, I thought about icon for completed item that's why I wrote about some refactoring. But actually, I don't think it's important enough to make a lot of changes so let's use parent icon for now